_posts/2020-10-23-deconvolution.md

```diff
@@ -0,0 +1,446 @@
+[Scikit-Image](https://github.com/scikit-image/scikit-image/blob/master/skimage/restoration/deconvolution.py#L329).
+We left off at theorizing how GPUs could potentially help accelerate these
+workflows.  Starting with Scikit-Image's implementation, we naively tried
+replacing `scipy.signal.convolve` calls with `cupyx.scipy.ndimage.convolve`,
+and while performance improved, it did not improve _significantly_ -- that is,
+we did not get the 100X speed we were looking for.
+
+As it often turns out in mathematics a problem that can be inefficient to solve
+in one representation can often be made more efficent by transforming the data
+beforehand. In this new representation we can solve the same problem
+(convolution in this case) more easily before transforming the result back into
+a more familiar representation. When it comes to convolution, the
+transformation we apply is called [Fast-Fourirer Transform
+(FFT)](https://en.wikipedia.org/wiki/Fast_Fourier_transform). Once this is
+applied we are able to convolve data using a simple multiplication.
+
+As it turns out this FFT transformation is extremely fast on both CPUs and
+GPUs. Similarly the algorithm we can write with FFTs are accelerated.  Despite
+the added step of doing FFTs, the cost of transformation + the cost of the
+algorithm is still lower than performing the original algorithm in real space.
+We (and others before us) found this was the case for Richardson Lucy (on both
+CPUs and GPUs) and performance continued increasing when we parallelized with
+Dask over multiple GPUs.
+
+Help from Open-Source
+---------------------
+
+An FFT RL equivalent has been around for some time and the good folks at the
+[Solar Dynamics Observatory](https://sdo.gsfc.nasa.gov/mission/instruments.php)
+built and shared a NumPy/CuPy implementation as part the [Atmospheric Imaging
+Assembly](https://aiapy.readthedocs.io/en/v0.2.0/_modules/aiapy/psf/deconvolve.html)
+Python package (aiapy).  We slightly modified their implementation to handle 3D
+as well as 2D [Point Spread
+Functions](https://en.wikipedia.org/wiki/Point_spread_function) and to take
+advantage of
+[NEP-18](https://numpy.org/neps/nep-0018-array-function-protocol.html) for
+convenient dispatching of NumPy and CuPy to NumPy and CuPy functions:
+
+
+```python
+def deconvolve(img, psf=None, iterations=20):
+    """
+    """
+    # 2D vs 3D psfs
+    if len(psf.shape) == 2:
+        psf = psf
+    else:
+        psf = psf[0]
+
+    # pad psf with zeros to match image shape
+    # use NumPy for padding
+    pad_l, pad_r = np.divmod(np.array(img[0].shape) - np.array(psf.shape), 2)
+    pad_r += pad_l
+
+    psf = np.pad(psf, tuple(zip(pad_l, pad_r)), 'constant', constant_values=0)
+
+    # padding centers the PSF so we no longer need to roll
+    # Center PSF at pixel (0,0)
+    for i in range(psf.ndim):
+        psf = np.roll(psf, psf.shape[i]//2, axis=i)
+
+    # Convolution requires FFT of the PSF
+    psf = np.fft.rfftn(psf)
+    psf = psf[None]
```

Comment:
What does this indexing do? I'm not familiar with the details of `rfftn`. A comment might help here.

Reply:
This adds a singleton dimension. That said, I'm not sure we still need this. Will recheck. Thanks!

Reply:
I think you should be able to remove the calls to `np.pad` and `np.roll` and just directly call rfftn with the `s` (shape) argument in rfftn:

```Python
psf = np.fft.rfftn(psf, s=img[0].shape)
```
In general indexing with `img[0]` and then adding a dimension back with `psf[None]` seems to assume a 2D PDF

Reply:
What do you think about leaving them in from an education standpoint? It seems useful for people to see the steps explicitly. Maybe we can make a note about using `s=`?

Yeah this code has been removed in a more recent rewrite. Please let us know if anything else sticks out :)

Reply:
Does `np.fft.rfftn(psf, s=img.shape)` also recenter the PSF around the origin? Or does one still need to multiply by a phase offset in FFT space to fix that? Just using `s` seems to move the deconvolved image somehow.